File: src/rnn/RNNTrainingSystem.cc
#include "rnn/RNNTrainingSystem.hh"

#include "debug/RNNTrainingSystem.hh"
#include "sim/sim_exit.hh"

namespace gem5 {

RNNTrainingSystem::RNNTrainingSystem(const Params &params)
    : ClockedObject(params),
      rnn(params.rnn),
      trainingData(params.training_data),
      targetOutputs(params.target_outputs),
      currentEpoch(0),
      maxEpochs(params.max_epochs) {
    // Schedule the first tick
    schedule(new EventFunctionWrapper([this] { tick(); }, name(), true),
    curTick() + clockPeriod());
}

void RNNTrainingSystem::tick() {
    DPRINTF(RNNTrainingSystem, "RNN Training System tick at %llu\n",
    curTick());

    if (currentEpoch < maxEpochs) {
        trainEpoch();
        currentEpoch++;
        // Schedule the next tick
        schedule(new EventFunctionWrapper([this] { tick(); }, name(), true),
        curTick() + clockPeriod());
    } else {
        DPRINTF(RNNTrainingSystem,
            "RNN Training completed after %d epochs\n", maxEpochs);
        exitSimLoop("RNN Training completed");
    }
}

void RNNTrainingSystem::trainEpoch() {
    DPRINTF(RNNTrainingSystem, "Training epoch %d\n", currentEpoch);

    for (size_t i = 0; i < trainingData.size(); i++) {
        rnn->processInput(trainingData[i]);
        std::vector<float> output = rnn->getOutput();
        // Backpropagate and update weights
        rnn->backpropagate(targetOutputs[i]);
        // Print the output
        DPRINTF(RNNTrainingSystem, "Sample %zu - Output: ", i);
        for (float val : output) {
            DPRINTF(RNNTrainingSystem, "%f ", val);
        }
        DPRINTF(RNNTrainingSystem, "\n");
    }
}

} // namespace gem5

File: rnn/RNNTrainingSystem.py
from m5.objects.ClockedObject import ClockedObject
from m5.params import *


class RNNTrainingSystem(ClockedObject):
    type = "RNNTrainingSystem"
    cxx_header = "rnn/RNNTrainingSystem.hh"
    cxx_class = "gem5::RNNTrainingSystem"

    rnn = Param.RNN("The RNN to train")

    # Define training_data as a vector of vectors of floats
    training_data = Param.VectorVectorFloat([], "Training data")

    # Define target_outputs as a vector of vectors of floats
    target_outputs = Param.VectorVectorFloat([], "Target outputs")

    max_epochs = Param.Int(1, "Maximum number of training epochs")
